### PR TITLE
[10.0][IMP] medical: Relicense and Improve

### PR DESCRIPTION
**Pull request by lasley**
This makes some updates to the core medical module that we have noticed now that we've been deployed with this version for a while. It also updates the licensing to LGPL. Detailed changelog:

* Relicense to LGPL (approval tally [here](https://github.com/OCA/vertical-medical/pull/166#issuecomment-280702415))
* Fix bugs identified after production deploys
* Add age search for patients
* Unify security
* Fix some code style issues
* Update abstract entity view for more modularity
* Remove AGPL assets and define necessary fields to match the naming for ecosystem compatibility
* Move patient medical information to notebook page
* Add patient smart button to partner

cc @LasLabs @Eficent

### CODE REVIEW

**Comment by TheCloneMaster** [comment]
Hi Dave,

First, thanks for your hard work on these modules. Great work!!

Now, I am not a layer, but...  can this license change be done just like
that?

Medical modules have several colaborators, starting from the original
creators, so as far as I know them all have to agree with this change, or
their work has to be rewritten ...

What is the purpose of this license change? What is it fixing that couldnt
on AGPL?

Regards,
Mario


El 16 may. 2017 7:27 PM, "Dave Lasley" <notifications@github.com> escribió:

This makes some updates to the core medical module that we have noticed now
that we've been deployed with this version for a while. It also updates the
licensing to LGPL. Detailed changelog:

   - Relicense to LGPL
   - Fix bugs identified after production deploys
   - Add age search for patients
   - Unify security
   - Fix some code style issues
   - Update abstract entity view for more modularity
   - Remove AGPL assets and define necessary fields to match the naming for
   ecosystem compatibility
   - Move patient medical information to notebook page
   - Add patient smart button to partner

cc @LasLabs <https://github.com/laslabs> @Eficent
<https://github.com/eficent>
------------------------------
You can view, comment on, or merge this pull request online at:

  https://github.com/OCA/vertical-medical/pull/173
Commit Summary

   - [IMP] medical: Relicense and Improve

File Changes

   - *M* medical/README.rst
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-0> (9)
   - *M* medical/__init__.py
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-1> (2)
   - *M* medical/__manifest__.py
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-2> (16)
   - *D* medical/i18n/ro.po
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-3> (960)
   - *M* medical/models/__init__.py
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-4> (2)
   - *M* medical/models/medical_abstract_entity.py
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-5> (58)
   - *M* medical/models/medical_patient.py
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-6> (57)
   - *M* medical/models/res_partner.py
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-7> (36)
   - *M* medical/security/ir.model.access.csv
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-8> (10)
   - *M* medical/security/medical_security.xml
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-9> (26)
   - *M* medical/templates/assets.xml
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-10> (2)
   - *M* medical/tests/test_medical_patient.py
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-11> (37)
   - *M* medical/views/medical_abstract_entity.xml
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-12> (19)
   - *M* medical/views/medical_menu.xml
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-13> (10)
   - *M* medical/views/medical_patient.xml
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-14> (58)
   - *M* medical/views/res_partner.xml
   <https://github.com/OCA/vertical-medical/pull/173/files#diff-15> (18)

Patch Links:

   - https://github.com/OCA/vertical-medical/pull/173.patch
   - https://github.com/OCA/vertical-medical/pull/173.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/OCA/vertical-medical/pull/173>, or mute the thread

**Comment by lasley** [comment]
@TheCloneMaster - Oh crap sorry I should have linked the [history of the license change](https://github.com/OCA/vertical-medical/pull/166). I've updated the PR comment with this too.

A license change most certainly cannot be made on the fly, but I've already made my case and obtained approvals. We just have finally gotten around to releasing the v10s since this conversation took place.

Once this and #174 are merged, I will make another PR for just the license change on the rest of the items. It had to be done this way due to the pre-existing `medical` v10 merge.

**Comment by TheCloneMaster** [comment]
Great!

Just wanted to be sure this will not be a problem later on...

Works for me...

Regards,
Mario

**Review by JordiBForgeFlow — APPROVED** [review]
:+1: Technical and functional review.

**Comment by lasley** [comment]
@obulkin please review

File: medical/models/medical_abstract_entity.py
Comment (obulkin) [inline medical/models/medical_abstract_entity.py]: This example needs to be updated. It's still using the old method name, and the following doesn't seem right given that the base version of the method always returns `None`:

```
if not res:
    return res
```

File: medical/models/medical_patient.py
Comment (obulkin) [inline medical/models/medical_patient.py]: What's the point of this line? Doesn't the conditional right above it guarantee that the operator will be in this list?

File: medical/models/medical_patient.py
Comment (obulkin) [inline medical/models/medical_patient.py]: In my mind, a baby has an age of 1 on its first birthday, rather than the day after, which makes these dates off by one day. Specifically, the last possible birthdate should be on the current day, while the first possible date should be one day ahead of the current day.

File: medical/models/medical_patient.py
Comment (obulkin) [inline medical/models/medical_patient.py]: The operators need to be reversed in all of these unequal cases, and the cutoff dates should be different. For example, if we want someone with an age `>=` a certain value, then the domain should be `[('birthdate_date', '<=', last_possible_birthdate)]`.

**Review by obulkin — CHANGES_REQUESTED** [review]
I'm seeing a few issues here, mostly in the logic for searching by age.

File: medical/models/medical_patient.py
Comment (gmeficent) [inline medical/models/medical_patient.py:120]: IMO, this treatment is confusing since it leads to incongruence in the user interface. See ex. below:

image

IMO, since the field Age is treated as a character the user can't choose 'greater/less than' operators anyway to create the filter and the only operator is going to be 'like' it would be better to leave the search behaviour as a string and let the user guess that he can use 'contains'-> '67y'. Also, this way allows to do month searches which would be useful for babies.

Maybe it would be useful a related, invisible, integer 'age2' field to add in the search view and allow comparatives? I don't know if operators can be added in the custom filters.

**Review by gmeficent — CHANGES_REQUESTED** [review]
Code and functional review :+1: 

Besides the comment below the rest LGTM

**Comment by JordiBForgeFlow** [comment]
I introduced fixes to the age search here: https://github.com/LasLabs/vertical-medical/pull/209

**Comment by jbeficent** [comment]
@Eficent/eficent

**Comment by jbeficent** [comment]
This is a test @Eficent/eficent

**Comment by lasley** [comment]
Thanks for the reviews @gmeficent & @obulkin - all comments attended to, primarily by @jbeficent's 37626c0

**Comment by obulkin** [comment]
@lasley - [This comment](https://github.com/OCA/vertical-medical/pull/173#discussion_r117859754) hasn't been addressed yet. Also, it occurs to me that the use of `replace` there will generate exceptions when the current day is the first day of the month (or the last day of the month after my proposed change).

**Comment by lasley** [comment]
@obulkin - I'm pretty sure 37626c0868ff78b4bc2b40bc5310c94e7227557c attended to https://github.com/OCA/vertical-medical/pull/173#discussion_r117859754, unless I'm reading this incorrectly. 

You do have a point with replace though I think.

**Comment by obulkin** [comment]
@lasley - I don't think it has. The logic in `_search_age` still has the issue of being off by one day (as well as the problems with `replace`). That commit took care of my other comments about the unnecessary `assert` and the reversed operators in all of the unequal domains.

**Comment by obulkin** [comment]
@lasley - I've pushed up a new commit that should address these issues and makes the branch ready for a merge, as far as I can tell.

**Review by obulkin — APPROVED** [review]

**Comment by obulkin** [comment]
@gmeficent - Could you take another look and approve the PR if your comments have been addressed? Thanks!

**Review by gmeficent — APPROVED** [review]
@obulkin, it has been solved :+1:

**Pull request merged**
